src/Sarif/Map/JsonMapBuilder.cs

```diff
@@ -15,58 +15,46 @@ namespace Microsoft.CodeAnalysis.Sarif.Map
-        private const long ArrayStartSizeEstimateBytes = 5;    // Bytes for one ArrayStart location, if under 10KB.
-
-        public double MaxFileSizePercentage { get; private set; }
-        public int MinimumSizeForNode { get; private set; }
-
-        /// <summary>
-        ///  Construct a JsonMapBuilder for the given target output size.
-        ///  (ex: 0.01 means 1% of the input file size)
-        /// </summary>
-        /// <param name="maxFileSizePercentage">Maximum size of the map relative to the source file (0.10 means 10%)</param>
-        public JsonMapBuilder(double maxFileSizePercentage)
-        {
-            if (maxFileSizePercentage <= 0 || maxFileSizePercentage >= 100) { throw new ArgumentOutOfRangeException("maxFileSizePercentage must be between 0 and 100."); }
-            MaxFileSizePercentage = maxFileSizePercentage;
-            MinimumSizeForNode = (int)(NodeSizeEstimateBytes / maxFileSizePercentage);
-        }
-
         /// <summary>
         ///  Build the JsonMap for a given Json file, by path.
         ///  Returns null if the source file was too small for any map nodes to fit the size budget.
         /// </summary>
         /// <param name="filePath">File Path to the Json file to build a map from</param>
+        /// <param name="settings">JsonMapSettings for map; null for defaults</param>
         /// <returns>JsonMap for file or null if file too small for map</returns>
-        public JsonMapNode Build(string filePath)
+        public static JsonMapNode Build(string filePath, JsonMapSettings settings = null)
         {
-            return Build(() => File.OpenRead(filePath));
+            return Build(() => File.OpenRead(filePath), settings);
         }
 
         /// <summary>
         ///  Build the JsonMap for a given Json file, given a stream provider.
         ///  Returns null if the source file was too small for any map nodes to fit the size budget.
         /// </summary>
         /// <param name="streamProvider">A function which will open the stream for the desired file</param>
+        /// <param name="settings">JsonMapSettings for map; null for defaults</param>
         /// <returns>JsonMap for file or null if file too small for map</returns>
-        public JsonMapNode Build(Func<Stream> streamProvider)
+        public static JsonMapNode Build(Func<Stream> streamProvider, JsonMapSettings settings = null)
         {
+            if (settings == null) { settings = JsonMapSettings.DefaultSettings; }
+
             using (Stream stream = streamProvider())
             {
-                if (stream.Length <= MinimumSizeForNode) { return null; }
+                long length = stream.Length;
+                if (length <= settings.MinimumSizeForNode) { return null; }
+                settings.AdjustForFileSize(length);
             }
 
             using (JsonPositionedTextReader reader = new JsonPositionedTextReader(streamProvider))
```

Comment:
>streamProvider [](start = 82, length = 14)

... and then I guess I have the same question about `JsonPositionedTextReader`; why doesn't it just take a `Stream`? ...

... oh. Is the problem that if the `JsonMapBuilder` and the `JsonPositionedTextReader` both use the same stream, that one of them (probably the reader) will dispose the stream before the other one (probably the builder) is done with it?

If that's the rationale, it needs a big old comment.

Reply:
The "big old comment" is present as a class comment on JsonPositionedTextReader. I've added more to explain why streamProvider() is needed. I've added a comment in SarifLog.LoadDeferred to explain both uses. I've added a comment in JsonMapBuilder.Build to explain why JsonPositionedTextReader is needed but not streamProvider (as it's not needed for this scenario but demanded by JsonPositionedTextReader for other scenarios). A confused reader who goes to the constructor to see what overloads are available will hopefully see the class comment where it is explained.